src/internet_identity/src/main.rs

```diff
@@ -477,6 +482,78 @@ async fn add(user_number: UserNumber, device_data: DeviceData) {
     })
 }
 
+/// Replace or remove an existing device.
+///
+/// NOTE: all mutable operations should call this function because it deals with protected stuff
+fn mutate_device_or_trap(
+    entries: &mut Vec<DeviceDataInternal>,
+    device_key: DeviceKey,
+    new_value: Option<DeviceData>,
+) {
+    let index = match entries.iter().position(|e| e.pubkey == device_key) {
+        None => trap("aie aie aie"),
+        Some(index) => index,
+    };
+
+    let e = entries.get_mut(index).unwrap();
+
+    // Run appropriate checks for protected devices
+    match e.protection_type {
+        None => (),
+        Some(ProtectionType::Unprotected) => (),
+        Some(ProtectionType::Protected) => {
+            // If the call is not authenticated with the device to mutate, abort
+            if caller() != Principal::self_authenticating(&e.pubkey) {
+                trap("Must be authenticated with protected device to mutate");
+            }
+        }
+    };
+
+    match new_value {
+        Some(device_data) => {
+            *e = device_data.into();
+        }
+        None => {
+            // NOTE: we void the more efficient remove_swap to ensure device ordering
```

Comment:
Is device ordering an important property?

Reply:
At the very least, performance is not really a concern here so might as well keep the ordering...